Gradle Inspector terminates with OutOfMemory

### Describe the bug

Since Ort versions 32.x and later, Gradle Inspector start to explode with Out of Memory crash.
Same unmodified code that worked with Ort 30.0.0.

### To Reproduce

Download [Demo Test tarball](https://dev.heliocastro.info/demo.tar.xz) and run `ort analyze -i . -o output`

### Expected behavior

Analyzer file to be created

### Console / log output

Add console and / or log output that shows the error and additional context.
No screenshots of plain text please, to keep text searchable.

```
 ______________________________                                                            
/        \_______   \__    ___/       The OSS Review Toolkit, version 33.1.0,              
|    |   | |       _/ |    |          built with JDK 11.0.24+8, running under Java 17.0.12.
|    |   | |    |   \ |    |          Executing 'analyze' as '***' on Mac OS X         
\________/ |____|___/ |____|          with 10 CPUs and a maximum of 8192 MiB of memory.    
                                                                                           
Environment variables:                                                                    
ORT_CONFIG_DIR = /Users/*****/data/ort-config                                           
ORT_DATA_DIR = /Users/****/.ort                                                        
HOME = /Users/******                                                                     
SHELL = /bin/zsh                                                                          
TERM = xterm-kitty                                                                        
http_proxy = *****
https_proxy = *****
JAVA_HOME = /opt/homebrew/opt/openjdk@17/libexec/openjdk.jdk/Contents/Home                
ANDROID_HOME = /Users/*****/.local/toolchains/android/android-sdk                       
                                                                                          
Looking for ORT configuration in the following file:
        /Users/*****/data/ort/config_host.yml

Looking for analyzer-specific configuration in the following files and directories:
        /Users/*****/code/testcode/.ort.yml
        /Users/****/data/ort-config/resolutions.yml
The following 1 package manager(s) are enabled:
        GradleInspector
The following 3 package curation provider(s) are enabled:
        RepositoryConfiguration, DefaultFile, DefaultDir
Analyzing project path:
        /Users/****/code/testcode
12:34:45.884 [main] INFO  org.ossreviewtoolkit.analyzer.PackageManager - Not analyzing directory '/Users/****/code/testcode/.gradle' as it is hard-coded to be ignored.
12:34:45.908 [main] INFO  org.ossreviewtoolkit.analyzer.PackageManager - Not analyzing directory '/Users/****/code/testcode/.git' as it is hard-coded to be ignored.
12:34:45.909 [main] INFO  org.ossreviewtoolkit.analyzer.PackageManager - Not analyzing directory '/Users/****/code/testcode/src/test/resources' as it is hard-coded to be ignored.
12:34:45.911 [main] INFO  org.ossreviewtoolkit.analyzer.PackageManager - Not analyzing directory '/Users/****/code/testcode/src/main/resources' as it is hard-coded to be ignored.
Found 1 GradleInspector definition file(s) at:
        build.gradle
Found in total 1 definition file(s) from the following 1 package manager(s):
        GradleInspector
12:34:45.941 [main] INFO  org.ossreviewtoolkit.analyzer.PackageManagerRunner - Starting GradleInspector analysis.
12:34:45.944 [DefaultDispatcher-worker-1] INFO  org.ossreviewtoolkit.analyzer.PackageManager - Using GradleInspector to resolve dependencies for path 'build.gradle'...
12:34:46.347 [DefaultDispatcher-worker-1] INFO  org.gradle.tooling.GradleConnector - The project at '/Users/DHXBWM7/code/cariad/user-push-service' uses Gradle version 8.5.0.
12:34:52.551 [DefaultDispatcher-worker-1] INFO  org.ossreviewtoolkit.utils.ort.OrtProxySelector - Proxy selector was successfully installed.
12:34:52.559 [DefaultDispatcher-worker-1] INFO  org.ossreviewtoolkit.utils.ort.OrtAuthenticator - Authenticator was successfully installed.
12:34:52.747 [DefaultDispatcher-worker-1] ERROR org.ossreviewtoolkit.analyzer.PackageManager - GradleInspector failed to resolve dependencies for path 'build.gradle': BuildException: Could not fetch model of type 'OrtDependencyTreeModel' using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-8.5-bin.zip'.
Caused by: LocationAwareException: Java heap space
    Caused by: OutOfMemoryError: Java heap space
12:34:52.749 [DefaultDispatcher-worker-1] INFO  org.ossreviewtoolkit.analyzer.PackageManager - GradleInspector resolved dependencies for path 'build.gradle' in 6.802362833s.
12:34:52.752 [DefaultDispatcher-worker-1] INFO  kotlinx.coroutines.CoroutineScope - Finished GradleInspector analysis.
12:34:52.787 [main] INFO  org.ossreviewtoolkit.utils.config.ConfigurationResolver - Getting 0 package curation(s) from provider 'RepositoryConfiguration' took 117.834us.
12:34:52.787 [main] INFO  org.ossreviewtoolkit.utils.config.ConfigurationResolver - Getting 0 package curation(s) from provider 'DefaultFile' took 1.167us.
12:34:52.788 [main] INFO  org.ossreviewtoolkit.utils.config.ConfigurationResolver - Getting 0 package curation(s) from provider 'DefaultDir' took 457.25us.
Wrote analyzer result to '/Users/****/code/testcode/output/analyzer-result.yml' (0,00 MiB) in 306.559417ms.
The analysis took 6.859496s.
Found 1 project(s) and 0 package(s) in total (not counting excluded ones).
Applied 0 curation(s) from 0 of 3 provider(s).
Resolved issues: 0 errors, 0 warnings, 0 hints.
Unresolved issues: 1 error, 0 warnings, 0 hints.
There is 1 unresolved issue with a severity equal to or greater than the ERROR threshold.
```

*Or* manually specify:

* ORT version: 33.1.0
* Java version: 17.0.12
* OS: MacOS

### Additional context

Add any other context about the problem here.

Tried with Java 21, and with extended 16Gb memory option for both JAVA_OPTS and GRADLE_OPTS

I got a second big internal Android project with monorepo, that was fine before, now some of the entries are issuing the same error:
```
07:04:09.939 [DefaultDispatcher-worker-1] ERROR org.ossreviewtoolkit.analyzer.PackageManager - GradleInspector failed to resolve dependencies for path 'cat-nav/ui-vw/build.gradle.kts': BuildException: Could not fetch model of type 'OrtDependencyTreeModel' using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-8.7-bin.zip'.
Caused by: LocationAwareException: Java heap space
    Caused by: OutOfMemoryError: Java heap space
```

@sschuberth revert temporary that commit you mentioned above is out of question ?

I have the same "Out of Memory" problems with a large Android Project, using _GradleInspector_ as Analyzer Plugin.
ORT Server with ORT 34.0.0

```
GradleInspector failed to resolve dependencies for path 'app/build.gradle': GradleConnectionException: Could not fetch model of type 'OrtDependencyTreeModel' using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-8.9-all.zip'.
Caused by: DaemonStoppedException: Gradle build daemon has been stopped: since the JVM garbage collector is thrashing
```

```
GradleInspector failed to resolve dependencies for path 'xxx/build.gradle': BuildException: Could not fetch model of type 'OrtDependencyTreeModel' using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-8.9-all.zip'.
Caused by: LocationAwareException: Java heap space
    Caused by: OutOfMemoryError: Java heap space
```

--- a/plugins/package-managers/gradle-plugin/src/main/kotlin/OrtModelBuilder.kt
+++ b/plugins/package-managers/gradle-plugin/src/main/kotlin/OrtModelBuilder.kt
@@ -54,6 +54,7 @@ internal class OrtModelBuilder : ToolingModelBuilder {
     private val logger = Logging.getLogger(OrtModelBuilder::class.java)
     private val errors = mutableListOf<String>()
     private val warnings = mutableListOf<String>()
+    private val globalDependencySubtrees = mutableMapOf<String, List<OrtDependency>>()
 
     override fun canBuild(modelName: String): Boolean = modelName == OrtDependencyTreeModel::class.java.name
 
@@ -205,7 +206,10 @@ internal class OrtModelBuilder : ToolingModelBuilder {
                                 warnings += message
                             }
 
-                            val dependencies = selectedComponent.dependencies.toOrtDependencies(poms, visited + id)
+                            // Check if we have scanned the dependencies of this subtree before, and if so, reuse them.
+                            val dependencies = globalDependencySubtrees.getOrPut(id.displayName) {
+                                selectedComponent.dependencies.toOrtDependencies(poms, visited + id)
+                            }
 
                             OrtDependencyImpl(
                                 groupId = id.group,

